# Spam Filter for Look2Social

## ~03/10

* Finished Collecting Data
* Communicated with Look2Social team: Figured out the goal of the program
* Communicated with instructors: Adjusted my approach to the project

## 03/14 ~ : Data Cleaning

### 03/14 - Overview of data
(screenshot of original data)
* 20,000 rows of data - not labeled for spam
* Six criteria for spam: 

    1) Marketing Focus (by social media marketing group)
    
    2) Bot Generated
    
    3) Known Spammer
    
    4) Corporate Posted
    
    5) Own Post
    
    6) Hijacking (just use tag, different content)


Bot Generated - can be labeled by either pattern of post ("just closed a deal...") and url

Corporate Posted - User input account name

Known spammer, Own Post - User input account name

Hijacking, Marketing - Need some eyeballing.

In [1]:
import pandas as pd
import numpy as np

### Suggestion from Miles

(pic of Miles)

"Use Machine Learning form early phase! Use it for labeling!" 

### Approach
1. create 7 columns for spam - one column per each criteria 
    one for final spam column that's default value is 0 but become 1 based on user declaration of spam

2. label bot generated and corporate generated first by their pattern

3. seperate out the rows with no label. (~7000 rows)

4. label first dozen~hundred for hijacking and marketing

5. use machine learning (Naive Bayes / Logistic Regression / Random Forest) to label, set up threshold high 

6. eyeball rest of 'em

### 03/17 - Steps before labeling data
Created new columns and index

In [2]:
df = pd.read_excel('data/data_prepping_zeros_index.xlsx')

In [7]:
df

index                                               text  \
0          0  just closed a deal in 27 hours using #AdobeSig...   
1          1  just closed a deal in 2 days using #AdobeSign ...   
2          2  just closed a deal in 2 hours using #AdobeSign...   
3          3  just closed a deal in 26 hours using #AdobeSig...   
4          4  just closed a deal in 6 days using #AdobeSign ...   
5          5  just closed a deal in 35 minutes using #AdobeS...   
6          6  just closed a deal in 3 days using #AdobeSign ...   
7          7  just closed a deal in 58 minutes using #AdobeS...   
8          8  just closed a deal in 7 minutes using #AdobeSi...   
9          9  So apparently IU is catching on and won’t ju...   
10        10  just closed a deal in 27 minutes using #AdobeS...   
11        11  just closed a deal in 21 minutes using #AdobeS...   
12        12  just closed a deal in 4 hours using #AdobeSign...   
13        13  just closed a deal in 28 minutes using #AdobeS...   
14        14  just closed a deal using #AdobeSign https://t....   
15        15  just closed a deal in 25 hours using #AdobeSig...   
16        16  just closed a deal in 3 hours using #AdobeSign...   
17        17  just closed a deal in 8 minutes using #AdobeSi...   
18        18  just closed a deal in 88 minutes using #AdobeS...   
19        19  just closed a deal in 29 hours using #AdobeSig...   
20        20  just closed a deal in 18 minutes using #AdobeS...   
21        21  just closed a deal in 5 minutes using #AdobeSi...   
22        22  just signed an agreement using #AdobeSign http...   
23        23  just closed a deal in 3 minutes using #AdobeSi...   
24        24  just closed a deal in 75 minutes using #AdobeS...   
25        25  just closed a deal in 4 minutes using #AdobeSi...   
26        26  just closed a deal in 12 minutes using #AdobeS...   
27        27  just closed a deal in 73 minutes using #AdobeS...   
28        28  just closed a deal in 3 hours using #AdobeSign...   
29        29  just closed a deal in 5 days using #AdobeSign ...   
...      ...                                                ...   
19981  19981  SignNow rescues your company from the unnecess...   
19982  19982  If you send out 1000 docs for HR/Comp signing,...   
19983  19983  Extend that last mile with a one-way or two-wa...   
19984  19984  SignNow helps put the puzzle pieces together f...   
19985  19985  With #SignNow, you are flexible enough to chan...   
19986  19986  Do you want a digital interactive &amp; engagi...   
19987  19987  @atharkhan @HelloSign @G2Crowd @RightSignature...   
19988  19988  Are you looking to make ACH Payments using Str...   
19989  19989  5 major benefits of choosing SignNow for #medi...   
19990  19990  Do you need to send a document to a group of p...   
19991  19991  @beezy It should be back up. It was down for a...   
19992  19992  #Didyouknow that you can create a link to put ...   
19993  19993  SignNow helps get patient forms signed faster ...   
19994  19994  You probably knew that you can save money with...   
19995  19995  Need to sign last minute #legal forms? SignNow...   
19996  19996  @claaack Hi Mr. November, I apologize. We are ...   
19997  19997  and sign the petition as it needs lots &amp; l...   
19998  19998  Using an Electronic Signature or E-Signature a...   
19999  19999  #LaborDay Weekend yet? Don't let work hold you...   
20000  20000  Are you having an event for your office for th...   
20001  20001  Summer is fading but the dollars in your bank ...   
20002  20002  Do you need to sign documents in person on an ...   
20003  20003  Are you stressed about getting a document sign...   
20004  20004  Please join in the effort to stop genocide in ...   
20005  20005  @Anafa It should be back up. It was down for a...   
20006  20006  @signnow Yes, they apologized for the issues b...   
20007  20007  SignNow ensures that you cannot be breached by...   
20008  20008  @claaack Hi Mr. November, Were you able 

1. Corporate post

Criteria for corporate = client(DocuSign) and its competitors (OneSpan, SignNow, Adobe Sign)

Explored 'screen_name' column and came up with following list of corporate account name:

In [3]:
adobe = ['Acrobat','Acrobat_GU','adobe_mabhatia','adobe1234567','AdobeCare','AdobeDocCloud','AdobeExpCloud','AdobeGov','adobemax','AdobeNews','AdobePartner','adobesignstx','AdobeStarProps','AdobeUK']
onespan = ['atOneSpan','OneSpan','OneSpanSign']
docusign = ['DocuSign','DocuSignAPAC','DocuSignAPI','DocuSignIMPACT','DocuSignUK','DocuSignING']
signnow = ['signnow']

From this observation, I found that it would be possible to label for spam_corporate based on that whether screen name contains certain company name or not (would be more efficient)

* bonus found: Adobe leverages its branch companies
* Q: Do we need multiple accounts or one accoun for social media marketing?

In [4]:
combined_lst = adobe+onespan+docusign+signnow
#combined_lst

In [6]:
#df[df['spam_']]

In [7]:
#df.loc[df['screen_name'] in (combined_lst), 'spam_corporate'] = 1

In [8]:
#df.loc[df['c1'] == 'Value', 'c2'] = 10

In [5]:
df['spam_corporate'][df.screen_name.isin(combined_lst)] = 1

/home/daniel/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


2. Bot generated post

Bot generated post for this case is by AdobeSign, in format of "Just closed a deal in..." / "Just signed an agreement in.." / "just connected my...". its expanded url column value is always adobe sign. 

For right now, we can use the assumption that if expanded url column value is link to AdobeSign, then it is bot generated post. However, it would be possible to process text and compare its vector (ex: use clustering) to capture the group of bot generated post with its distinct pattern (later task to elaborate the work)

Considering time restriction, I will proceed with this hypothesis.

In [6]:
df['spam_bot'][df.expanded_url == 'https://acrobat.adobe.com'] = 1

/home/daniel/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [7]:
df.head()

index                                               text  retweet_count  \
0      0  just closed a deal in 27 hours using #AdobeSig...            NaN   
1      1  just closed a deal in 2 days using #AdobeSign ...            NaN   
2      2  just closed a deal in 2 hours using #AdobeSign...            NaN   
3      3  just closed a deal in 26 hours using #AdobeSig...            NaN   
4      4  just closed a deal in 6 days using #AdobeSign ...            NaN   

   favorite_count  spam  spam_marketing  spam_hijack  spam_corporate  \
0             NaN     0               0            0               0   
1             NaN     0               0            0               0   
2             NaN     0               0            0               0   
3             NaN     0               0            0               0   
4             NaN     0               0            0               0   

   spam_bot  spam_known  ...  place_type  full_name  place_name  place_id  \
0         1           0  ...         NaN        NaN         NaN       NaN   
1         1           0  ...         NaN        NaN         NaN       NaN   
2         1           0  ...         NaN        NaN         NaN       NaN   
3         1           0  ...         NaN        NaN         NaN       NaN   
4         1           0  ...         NaN        NaN         NaN       NaN   

   place_lat  place_lon  lat  lon               expanded_url  \
0        NaN        NaN  NaN  NaN  https://acrobat.adobe.com   
1        NaN        NaN  NaN  NaN  https://acrobat.adobe.com   
2        NaN        NaN  NaN  NaN  https://acrobat.adobe.com   
3        NaN        NaN  NaN  NaN  https://acrobat.adobe.com   
4        NaN        NaN  NaN  NaN  https://acrobat.adobe.com   

                       url  
0  https://t.co/Fe0YfarG31  
1  https://t.co/TTlCvaDE0V  
2  https://t.co/okmHDeM9bZ  
3  https://t.co/AAhUGDLZxH  
4  https://t.co/Oi57q6jdXP  

[5 rows x 55 columns]

* Q: is bot generated post worth to do it? Is that why Adobe using it?

Known spammer, Own Post spam would be labeled with similar method like below:

In [13]:
df.columns

Index(['index', 'text', 'retweet_count', 'favorite_count', 'spam',
       'spam_marketing', 'spam_hijack', 'spam_corporate', 'spam_bot',
       'spam_known', 'spam_own', 'Docusign', 'onespan', 'signnow',
       'adobe sign', 'favorited', 'truncated', 'id_str',
       'in_reply_to_screen_name', 'source', 'retweeted', 'created_at',
       'in_reply_to_status_id_str', 'in_reply_to_user_id_str', 'lang',
       'listed_count', 'verified', 'location', 'user_id_str', 'description',
       'geo_enabled', 'user_created_at', 'statuses_count', 'followers_count',
       'favourites_count', 'protected', 'user_url', 'name', 'time_zone',
       'user_lang', 'utc_offset', 'friends_count', 'screen_name',
       'country_code', 'country', 'place_type', 'full_name', 'place_name',
       'place_id', 'place_lat', 'place_lon', 'lat', 'lon', 'expanded_url',
       'url'],
      dtype='object')

And... finally eyeball time.
Let's seperate out the rows that none of spam categories value is 1.

In [14]:
df.loc[df['spam_bot'] == 1 | df['spam_corporate'] == 1]

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [14]:
df[(df['spam_bot'] == 1) | (df['spam_corporate'] == 1)]

index                                               text  \
0          0  just closed a deal in 27 hours using #AdobeSig...   
1          1  just closed a deal in 2 days using #AdobeSign ...   
2          2  just closed a deal in 2 hours using #AdobeSign...   
3          3  just closed a deal in 26 hours using #AdobeSig...   
4          4  just closed a deal in 6 days using #AdobeSign ...   
5          5  just closed a deal in 35 minutes using #AdobeS...   
6          6  just closed a deal in 3 days using #AdobeSign ...   
7          7  just closed a deal in 58 minutes using #AdobeS...   
8          8  just closed a deal in 7 minutes using #AdobeSi...   
10        10  just closed a deal in 27 minutes using #AdobeS...   
11        11  just closed a deal in 21 minutes using #AdobeS...   
12        12  just closed a deal in 4 hours using #AdobeSign...   
13        13  just closed a deal in 28 minutes using #AdobeS...   
14        14  just closed a deal using #AdobeSign https://t....   
15        15  just closed a deal in 25 hours using #AdobeSig...   
16        16  just closed a deal in 3 hours using #AdobeSign...   
17        17  just closed a deal in 8 minutes using #AdobeSi...   
18        18  just closed a deal in 88 minutes using #AdobeS...   
19        19  just closed a deal in 29 hours using #AdobeSig...   
20        20  just closed a deal in 18 minutes using #AdobeS...   
21        21  just closed a deal in 5 minutes using #AdobeSi...   
22        22  just signed an agreement using #AdobeSign http...   
23        23  just closed a deal in 3 minutes using #AdobeSi...   
24        24  just closed a deal in 75 minutes using #AdobeS...   
25        25  just closed a deal in 4 minutes using #AdobeSi...   
26        26  just closed a deal in 12 minutes using #AdobeS...   
27        27  just closed a deal in 73 minutes using #AdobeS...   
28        28  just closed a deal in 3 hours using #AdobeSign...   
29        29  just closed a deal in 5 days using #AdobeSign ...   
30        30  just closed a deal in 6 hours using #AdobeSign...   
...      ...                                                ...   
19970  19970  Are you a small business looking to streamline...   
19971  19971  What eSignature solution should you be signing...   
19972  19972  Did you know you can earn up to $300 when you ...   
19975  19975  Looking to streamline your document workflow? ...   
19976  19976  Technology is evolving in the #healthcare indu...   
19977  19977  We love referrals! Make extra money this #summ...   
19980  19980  Did you know health plans and #healthcare prov...   
19981  19981  SignNow rescues your company from the unnecess...   
19984  19984  SignNow helps put the puzzle pieces together f...   
19986  19986  Do you want a digital interactive &amp; engagi...   
19987  19987  @atharkhan @HelloSign @G2Crowd @RightSignature...   
19988  19988  Are you looking to make ACH Payments using Str...   
19989  19989  5 major benefits of choosing SignNow for #medi...   
19990  19990  Do you need to send a document to a group of p...   
19991  19991  @beezy It should be back up. It was down for a...   
19992  19992  #Didyouknow that you can create a link to put ...   
19993  19993  SignNow helps get patient forms signed faster ...   
19994  19994  You probably knew that you can save money with...   
19995  19995  Need to sign last minute #legal forms? SignNow...   
19996  19996  @claaack Hi Mr. November, I apologize. We are ...   
19999  19999  #LaborDay Weekend yet? Don't let work hold you...   
20000  20000  Are you having an event for your office for th...   
20001  20001  Summer is fading but the dollars in your bank ...   
20002  20002  Do you need to sign documents in person on an ...   
20003  20003  Are you stressed about getting a document sign...   
20005  20005  @Anafa It should be back up. It was down for a...   
20007  20007  SignNow ensures that you cannot be breached by...   
20008  20008  @claaack Hi Mr. November, Were you able 

In [15]:
target_df = df[(df['spam_bot'] != 1) & (df['spam_corporate'] != 1)]
target_df

index                                               text  \
9          9  So apparently IU is catching on and won’t ju...   
82        82  The 11 Year Customer: Last week, I met a VP of...   
107      107  7 Reasons to Choose Adobe Sign for E-Signature...   
115      115  What is it with ink discrimination? I was just...   
129      129  Explore Adobe Sign, a web base tool used for d...   
132      132  New version of Adobe Flash Player for Windows ...   
143      143  Adobe Fill &amp; Sign users, you can get aroun...   
151      151  Sign up for Coastline's Online DGA C164 Adobe ...   
163      163  Sapho Employee Experience Portal 4.9 is here! ...   
167      167  Sign up today for our summer 2018 classes! Top...   
186      186  Why won’t Adobe Illustrator the detect Phili...   
192      192  @HEAD_MASTA @its_kaymoney ‘Adobe Fill &amp; ...   
211      211  There's still time to sign up for @LarryJordan...   
238      238  Just signed up and completed a free #pd #cpd c...   
253      253  There's still time to sign up for @LarryJordan...   
256      256  @Bolaabae There’s a fill &amp; sign option o...   
274      274  Someone any one know how to sign documents on ...   
288      288  There's still time to sign up for @LarryJordan...   
289      289  Our new Wine Club Members Lounge will be open ...   
334      334  Starts Saturday. Learn how to edit using Adobe...   
343      343  Only 2 spots left for our INTRO TO ADOBE PREMI...   
347      347  @Adobe spent 1 hour today in the middle of bus...   
358      358  @PaulPerezInk @JonnyEthco @NeebsOfficial Do yo...   
369      369  @ZylphaLtd | The UK’s leading integration sp...   
405      405  This Wednesday's Larry Jordan #PowerUp Webinar...   
416      416  Now, at this point, I was subscribed to the Ad...   
451      451  being an art graduate is when you can't afford...   
458      458  This week's Larry Jordan #PowerUp Webinar is g...   
486      486  @ViktorFagerLFC Are we far enough down the lin...   
494      494  This Wednesday's Larry Jordan #PowerUp Webinar...   
...      ...                                                ...   
19923  19923  Cyber Times В® is out! https://t.co/p1kNm1zBiy...   
19926  19926  @HelloSign @G2Crowd I use #SignNow but would s...   
19929  19929  With #SignNow, you can see who is and who isn'...   
19931  19931  #SignNow has an audit trail for proactivity st...   
19933  19933  NEW RELEASE! Call in today to learn about Sign...   
19935  19935  @signnow @HelloSign @G2Crowd @RightSignature @...   
19941  19941  Background on ESIGN &amp; UETA https://t.co/sh...   
19947  19947  @signnow is there any trouble with your websit...   
19948  19948  For the first #SignNow pilot, we suggest getti...   
19950  19950  Enterprise Expansion Phase 3 out of 3: Now tha...   
19952  19952  #NetSuite helps automate the org using it to b...   
19953  19953  Be sure to stop by the office first thing in t...   
19954  19954  Select #SignNow pilots, a test group of 10-15 ...   
19955  19955  Hi @signnow I'm trying to get started on the p...   
19956  19956  When implementing #SignNow, templatize your pr...   
19958  19958  @cbledsoe7 heya! Been trying to get ahold of @...   
19959  19959  I liked a @YouTube video https://t.co/Vgbobxpc...   
19962  19962  You don't necessarily need IT if you want to f...   
19969  19969  #Breaking #Tell this Mayor John A McDonald mus...   
19973  19973  The #SignNow payoff is amplifying your #NetSui...   
19974  19974  SignNow lets you provide additional services i...   
19978  19978  @signnow When will your service be back up? Ge...   
19979  19979  #SignNow allows you to redline, edit, redact, ...   
19982  19982  If you send out 1000 docs for HR/Comp signing,...   
19983  19983  Extend that last mile with a one-way or two-wa...   
19985  19985  With #SignNow, you are flexible enough to chan...   
19997  19997  and sign the petition as it needs lots &amp; l...   
19998  19998  Using an Electronic Signature or E-Signa

In [18]:
13227+6784

20011

In [19]:
target_df = df[(df['spam_bot'] != 1) & (df['spam_corporate'] != 1)]

In [10]:
target_df.columns

Index(['index', 'text', 'retweet_count', 'favorite_count', 'spam',
       'spam_marketing', 'spam_hijack', 'spam_corporate', 'spam_bot',
       'spam_known', 'spam_own', 'Docusign', 'onespan', 'signnow',
       'adobe sign', 'favorited', 'truncated', 'id_str',
       'in_reply_to_screen_name', 'source', 'retweeted', 'created_at',
       'in_reply_to_status_id_str', 'in_reply_to_user_id_str', 'lang',
       'listed_count', 'verified', 'location', 'user_id_str', 'description',
       'geo_enabled', 'user_created_at', 'statuses_count', 'followers_count',
       'favourites_count', 'protected', 'user_url', 'name', 'time_zone',
       'user_lang', 'utc_offset', 'friends_count', 'screen_name',
       'country_code', 'country', 'place_type', 'full_name', 'place_name',
       'place_id', 'place_lat', 'place_lon', 'lat', 'lon', 'expanded_url',
       'url'],
      dtype='object')

In [13]:
target_df2 = target_df[['index','text','spam_marketing','spam_hijack','description','screen_name','expanded_url']]

In [14]:
target_df2.head()

index                                               text  spam_marketing  \
9        9  So apparently IU is catching on and won’t ju...               0   
82      82  The 11 Year Customer: Last week, I met a VP of...               0   
107    107  7 Reasons to Choose Adobe Sign for E-Signature...               0   
115    115  What is it with ink discrimination? I was just...               0   
129    129  Explore Adobe Sign, a web base tool used for d...               0   

     spam_hijack                                        description  \
9              0  BSU ‘17 - SIU ‘19 // Architecture & Interi...   
82             0  Changing the way management consulting service...   
107            0  Making your business STAND OUT in the digital ...   
115            0  Mythical conservative academic. Don't tell any...   
129            0  Cornell Information Technologies. Get support ...   

        screen_name                                       expanded_url  
9       cmdavis4cap                                                NaN  
82      MHPAdvisory                              http://dlvr.it/QWJnCz  
107  GleesonDigital                            https://buff.ly/2LZaGMH  
115  LadyGrammarian                                                NaN  
129      Cornell_IT  https://cornell.sabacloud.com/Saba/Web_spf/NA1...

In [1]:
#target_df2.to_excel("smalleyeball.xlsx", sheet_name='Sheet_name_1')

In [2]:
#target_df.to_excel("eyeball.xlsx", sheet_name='Sheet_name_1')

### 3/31 Finally done with labeling!

In [8]:
auto_labeled_df = df[(df['spam_bot'] == 1) | (df['spam_corporate'] == 1)]

In [9]:
hand_labeled_df = df[(df['spam_bot'] != 1) & (df['spam_corporate'] != 1)]

In [10]:
hand_labeled_df

index                                               text  \
9          9  So apparently IU is catching on and won’t ju...   
82        82  The 11 Year Customer: Last week, I met a VP of...   
107      107  7 Reasons to Choose Adobe Sign for E-Signature...   
115      115  What is it with ink discrimination? I was just...   
129      129  Explore Adobe Sign, a web base tool used for d...   
132      132  New version of Adobe Flash Player for Windows ...   
143      143  Adobe Fill &amp; Sign users, you can get aroun...   
151      151  Sign up for Coastline's Online DGA C164 Adobe ...   
163      163  Sapho Employee Experience Portal 4.9 is here! ...   
167      167  Sign up today for our summer 2018 classes! Top...   
186      186  Why won’t Adobe Illustrator the detect Phili...   
192      192  @HEAD_MASTA @its_kaymoney ‘Adobe Fill &amp; ...   
211      211  There's still time to sign up for @LarryJordan...   
238      238  Just signed up and completed a free #pd #cpd c...   
253      253  There's still time to sign up for @LarryJordan...   
256      256  @Bolaabae There’s a fill &amp; sign option o...   
274      274  Someone any one know how to sign documents on ...   
288      288  There's still time to sign up for @LarryJordan...   
289      289  Our new Wine Club Members Lounge will be open ...   
334      334  Starts Saturday. Learn how to edit using Adobe...   
343      343  Only 2 spots left for our INTRO TO ADOBE PREMI...   
347      347  @Adobe spent 1 hour today in the middle of bus...   
358      358  @PaulPerezInk @JonnyEthco @NeebsOfficial Do yo...   
369      369  @ZylphaLtd | The UK’s leading integration sp...   
405      405  This Wednesday's Larry Jordan #PowerUp Webinar...   
416      416  Now, at this point, I was subscribed to the Ad...   
451      451  being an art graduate is when you can't afford...   
458      458  This week's Larry Jordan #PowerUp Webinar is g...   
486      486  @ViktorFagerLFC Are we far enough down the lin...   
494      494  This Wednesday's Larry Jordan #PowerUp Webinar...   
...      ...                                                ...   
19923  19923  Cyber Times В® is out! https://t.co/p1kNm1zBiy...   
19926  19926  @HelloSign @G2Crowd I use #SignNow but would s...   
19929  19929  With #SignNow, you can see who is and who isn'...   
19931  19931  #SignNow has an audit trail for proactivity st...   
19933  19933  NEW RELEASE! Call in today to learn about Sign...   
19935  19935  @signnow @HelloSign @G2Crowd @RightSignature @...   
19941  19941  Background on ESIGN &amp; UETA https://t.co/sh...   
19947  19947  @signnow is there any trouble with your websit...   
19948  19948  For the first #SignNow pilot, we suggest getti...   
19950  19950  Enterprise Expansion Phase 3 out of 3: Now tha...   
19952  19952  #NetSuite helps automate the org using it to b...   
19953  19953  Be sure to stop by the office first thing in t...   
19954  19954  Select #SignNow pilots, a test group of 10-15 ...   
19955  19955  Hi @signnow I'm trying to get started on the p...   
19956  19956  When implementing #SignNow, templatize your pr...   
19958  19958  @cbledsoe7 heya! Been trying to get ahold of @...   
19959  19959  I liked a @YouTube video https://t.co/Vgbobxpc...   
19962  19962  You don't necessarily need IT if you want to f...   
19969  19969  #Breaking #Tell this Mayor John A McDonald mus...   
19973  19973  The #SignNow payoff is amplifying your #NetSui...   
19974  19974  SignNow lets you provide additional services i...   
19978  19978  @signnow When will your service be back up? Ge...   
19979  19979  #SignNow allows you to redline, edit, redact, ...   
19982  19982  If you send out 1000 docs for HR/Comp signing,...   
19983  19983  Extend that last mile with a one-way or two-wa...   
19985  19985  With #SignNow, you are flexible enough to chan...   
19997  19997  and sign the petition as it needs lots &amp; l...   
19998  19998  Using an Electronic Signature or E-Signa

In [20]:
13227+6784 #confirming

20011

Combining two subsets of hand-labeled data

In [11]:
hand_partone = pd.read_excel('data/smalleyeball_partone.xlsx')
hand_parttwo = pd.read_excel('data/smalleyeball_parttwo.xlsx')

In [51]:
#hand_partone

In [52]:
#hand_parttwo

In [12]:
hand_combined = pd.concat([hand_partone,hand_parttwo])

In [53]:
#hand_combined

In [54]:
#hand_combined.columns

In [13]:
hand_combined.drop(['Unnamed: 0', 'Unnamed: 0.1'], axis=1,inplace=True)

In [14]:
hand_combined.columns

Index(['index', 'text', 'spam_marketing', 'spam_hijack', 'description',
       'screen_name', 'expanded_url'],
      dtype='object')

In [55]:
#hand_combined

In [56]:
#hand_labeled_df

In [57]:
#print(hand_labeled_df.columns)

In [58]:
#print(hand_combined.columns)

array([0, 0, 1, ..., 1, 0, 0])

In [15]:
hand_labeled_df['spam_marketing'] = hand_combined['spam_marketing'].values

/home/daniel/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [16]:
hand_labeled_df['spam_hijack'] = hand_combined['spam_hijack'].values

/home/daniel/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [59]:
#hand_labeled_df

I don't have a confidence for consistency in data labeling & my understanding for the definition of spam marketing & hijack - it would be necessary for Look2Social to label once again with their standard.

It is clearly easier to label if you sort the file by post text - would be able to label repeating/similar post at once, which would save a lot of time.

There were so many insights from data labeling. I would personally recommend labeling data once a quarter / 2 quarters to understand the trend. It would capture some qualitative insights of the client(DocuSign) and its competitors (Adobe Sign, OneSpan, SignNow).

In [17]:
combined_labeled_df = pd.concat([hand_labeled_df,auto_labeled_df])

In [18]:
combined_labeled_df.sort_values(by='index', inplace=True)

In [19]:
combined_labeled_df

index                                               text  \
0          0  just closed a deal in 27 hours using #AdobeSig...   
1          1  just closed a deal in 2 days using #AdobeSign ...   
2          2  just closed a deal in 2 hours using #AdobeSign...   
3          3  just closed a deal in 26 hours using #AdobeSig...   
4          4  just closed a deal in 6 days using #AdobeSign ...   
5          5  just closed a deal in 35 minutes using #AdobeS...   
6          6  just closed a deal in 3 days using #AdobeSign ...   
7          7  just closed a deal in 58 minutes using #AdobeS...   
8          8  just closed a deal in 7 minutes using #AdobeSi...   
9          9  So apparently IU is catching on and won’t ju...   
10        10  just closed a deal in 27 minutes using #AdobeS...   
11        11  just closed a deal in 21 minutes using #AdobeS...   
12        12  just closed a deal in 4 hours using #AdobeSign...   
13        13  just closed a deal in 28 minutes using #AdobeS...   
14        14  just closed a deal using #AdobeSign https://t....   
15        15  just closed a deal in 25 hours using #AdobeSig...   
16        16  just closed a deal in 3 hours using #AdobeSign...   
17        17  just closed a deal in 8 minutes using #AdobeSi...   
18        18  just closed a deal in 88 minutes using #AdobeS...   
19        19  just closed a deal in 29 hours using #AdobeSig...   
20        20  just closed a deal in 18 minutes using #AdobeS...   
21        21  just closed a deal in 5 minutes using #AdobeSi...   
22        22  just signed an agreement using #AdobeSign http...   
23        23  just closed a deal in 3 minutes using #AdobeSi...   
24        24  just closed a deal in 75 minutes using #AdobeS...   
25        25  just closed a deal in 4 minutes using #AdobeSi...   
26        26  just closed a deal in 12 minutes using #AdobeS...   
27        27  just closed a deal in 73 minutes using #AdobeS...   
28        28  just closed a deal in 3 hours using #AdobeSign...   
29        29  just closed a deal in 5 days using #AdobeSign ...   
...      ...                                                ...   
19981  19981  SignNow rescues your company from the unnecess...   
19982  19982  If you send out 1000 docs for HR/Comp signing,...   
19983  19983  Extend that last mile with a one-way or two-wa...   
19984  19984  SignNow helps put the puzzle pieces together f...   
19985  19985  With #SignNow, you are flexible enough to chan...   
19986  19986  Do you want a digital interactive &amp; engagi...   
19987  19987  @atharkhan @HelloSign @G2Crowd @RightSignature...   
19988  19988  Are you looking to make ACH Payments using Str...   
19989  19989  5 major benefits of choosing SignNow for #medi...   
19990  19990  Do you need to send a document to a group of p...   
19991  19991  @beezy It should be back up. It was down for a...   
19992  19992  #Didyouknow that you can create a link to put ...   
19993  19993  SignNow helps get patient forms signed faster ...   
19994  19994  You probably knew that you can save money with...   
19995  19995  Need to sign last minute #legal forms? SignNow...   
19996  19996  @claaack Hi Mr. November, I apologize. We are ...   
19997  19997  and sign the petition as it needs lots &amp; l...   
19998  19998  Using an Electronic Signature or E-Signature a...   
19999  19999  #LaborDay Weekend yet? Don't let work hold you...   
20000  20000  Are you having an event for your office for th...   
20001  20001  Summer is fading but the dollars in your bank ...   
20002  20002  Do you need to sign documents in person on an ...   
20003  20003  Are you stressed about getting a document sign...   
20004  20004  Please join in the effort to stop genocide in ...   
20005  20005  @Anafa It should be back up. It was down for a...   
20006  20006  @signnow Yes, they apologized for the issues b...   
20007  20007  SignNow ensures that you cannot be breached by...   
20008  20008  @claaack Hi Mr. November, Were you able 

In [20]:
combined_labeled_df.columns

Index(['index', 'text', 'retweet_count', 'favorite_count', 'spam',
       'spam_marketing', 'spam_hijack', 'spam_corporate', 'spam_bot',
       'spam_known', 'spam_own', 'Docusign', 'onespan', 'signnow',
       'adobe sign', 'favorited', 'truncated', 'id_str',
       'in_reply_to_screen_name', 'source', 'retweeted', 'created_at',
       'in_reply_to_status_id_str', 'in_reply_to_user_id_str', 'lang',
       'listed_count', 'verified', 'location', 'user_id_str', 'description',
       'geo_enabled', 'user_created_at', 'statuses_count', 'followers_count',
       'favourites_count', 'protected', 'user_url', 'name', 'time_zone',
       'user_lang', 'utc_offset', 'friends_count', 'screen_name',
       'country_code', 'country', 'place_type', 'full_name', 'place_name',
       'place_id', 'place_lat', 'place_lon', 'lat', 'lon', 'expanded_url',
       'url'],
      dtype='object')

Let's cleaning data now. 

I will drop all column with only one unique value.

In [21]:
df2 = combined_labeled_df
for col in combined_labeled_df.columns:
    if col in ['spam', 'spam_own', 'spam_corporate', 'spam_known', 'spam_marketing', 'spam_hijack', 'spam_bot']:
        pass
    elif combined_labeled_df[col].nunique() == 1:
        df2 = df2.drop(col,axis=1)

In [22]:
df2.columns

Index(['index', 'text', 'retweet_count', 'favorite_count', 'spam',
       'spam_marketing', 'spam_hijack', 'spam_corporate', 'spam_bot',
       'spam_known', 'spam_own', 'Docusign', 'onespan', 'signnow',
       'adobe sign', 'id_str', 'in_reply_to_screen_name', 'source',
       'created_at', 'in_reply_to_status_id_str', 'in_reply_to_user_id_str',
       'lang', 'listed_count', 'verified', 'location', 'user_id_str',
       'description', 'geo_enabled', 'user_created_at', 'statuses_count',
       'followers_count', 'favourites_count', 'user_url', 'name', 'time_zone',
       'user_lang', 'utc_offset', 'friends_count', 'screen_name',
       'country_code', 'country', 'place_type', 'full_name', 'place_name',
       'place_id', 'place_lat', 'place_lon', 'lat', 'lon', 'expanded_url',
       'url'],
      dtype='object')

In [23]:
df2.head()

index                                               text  retweet_count  \
0      0  just closed a deal in 27 hours using #AdobeSig...            NaN   
1      1  just closed a deal in 2 days using #AdobeSign ...            NaN   
2      2  just closed a deal in 2 hours using #AdobeSign...            NaN   
3      3  just closed a deal in 26 hours using #AdobeSig...            NaN   
4      4  just closed a deal in 6 days using #AdobeSign ...            NaN   

   favorite_count  spam  spam_marketing  spam_hijack  spam_corporate  \
0             NaN     0               0            0               0   
1             NaN     0               0            0               0   
2             NaN     0               0            0               0   
3             NaN     0               0            0               0   
4             NaN     0               0            0               0   

   spam_bot  spam_known  ...  place_type  full_name  place_name  place_id  \
0         1           0  ...         NaN        NaN         NaN       NaN   
1         1           0  ...         NaN        NaN         NaN       NaN   
2         1           0  ...         NaN        NaN         NaN       NaN   
3         1           0  ...         NaN        NaN         NaN       NaN   
4         1           0  ...         NaN        NaN         NaN       NaN   

   place_lat  place_lon  lat lon               expanded_url  \
0        NaN        NaN  NaN NaN  https://acrobat.adobe.com   
1        NaN        NaN  NaN NaN  https://acrobat.adobe.com   
2        NaN        NaN  NaN NaN  https://acrobat.adobe.com   
3        NaN        NaN  NaN NaN  https://acrobat.adobe.com   
4        NaN        NaN  NaN NaN  https://acrobat.adobe.com   

                       url  
0  https://t.co/Fe0YfarG31  
1  https://t.co/TTlCvaDE0V  
2  https://t.co/okmHDeM9bZ  
3  https://t.co/AAhUGDLZxH  
4  https://t.co/Oi57q6jdXP  

[5 rows x 51 columns]

Good number of columns are filled with NaN.

I personally think it is danger to fill the column if NaN is more than half of row. I will use that as a threshold

In [24]:
df2.dropna(thresh=(len(df)/2), axis=1, inplace=True)

In [25]:
df2.head()

index                                               text  spam  \
0      0  just closed a deal in 27 hours using #AdobeSig...     0   
1      1  just closed a deal in 2 days using #AdobeSign ...     0   
2      2  just closed a deal in 2 hours using #AdobeSign...     0   
3      3  just closed a deal in 26 hours using #AdobeSig...     0   
4      4  just closed a deal in 6 days using #AdobeSign ...     0   

   spam_marketing  spam_hijack  spam_corporate  spam_bot  spam_known  \
0               0            0               0         1           0   
1               0            0               0         1           0   
2               0            0               0         1           0   
3               0            0               0         1           0   
4               0            0               0         1           0   

   spam_own  Docusign  ...                 user_created_at  statuses_count  \
0         0     False  ...  Fri Sep 09 20:40:00 +0000 2011            6750   
1         0     False  ...  Fri May 25 01:27:15 +0000 2012            5018   
2         0     False  ...  Thu Dec 18 22:06:40 +0000 2008              29   
3         0     False  ...  Wed Aug 08 20:11:12 +0000 2012            5332   
4         0     False  ...  Thu Aug 26 20:50:06 +0000 2010            1337   

   followers_count  favourites_count             name user_lang friends_count  \
0              108                 1      Joe Zarroli        en            12   
1               47                 2  Party Karacters        en            25   
2               60                10      annpodlozny        en            52   
3               66                 0     Angela Dugan        en           182   
4              225                19            Phebe        en           348   

       screen_name               expanded_url                      url  
0  IslandRealtyGrp  https://acrobat.adobe.com  https://t.co/Fe0YfarG31  
1   partykaracters  https://acrobat.adobe.com  https://t.co/TTlCvaDE0V  
2      annpodlozny  https://acrobat.adobe.com  https://t.co/okmHDeM9bZ  
3   AngelaRedondo1  https://acrobat.adobe.com  https://t.co/AAhUGDLZxH  
4   Makingmeevents  https://acrobat.adobe.com  https://t.co/Oi57q6jdXP  

[5 rows x 33 columns]

In [26]:
df2.columns

Index(['index', 'text', 'spam', 'spam_marketing', 'spam_hijack',
       'spam_corporate', 'spam_bot', 'spam_known', 'spam_own', 'Docusign',
       'onespan', 'signnow', 'adobe sign', 'id_str', 'source', 'created_at',
       'lang', 'listed_count', 'verified', 'location', 'user_id_str',
       'description', 'geo_enabled', 'user_created_at', 'statuses_count',
       'followers_count', 'favourites_count', 'name', 'user_lang',
       'friends_count', 'screen_name', 'expanded_url', 'url'],
      dtype='object')

In [32]:
df2[['id_str','source','created_at','lang','verified','user_id_str','geo_enabled','user_created_at']].head()

id_str                                             source  \
0  1004148045292889984  <a href="http://www.echosign.com/" rel="nofoll...   
1  1005207775352380032  <a href="http://www.echosign.com/" rel="nofoll...   
2  1005563350451929984  <a href="http://www.echosign.com/" rel="nofoll...   
3  1004856511322119936  <a href="http://www.echosign.com/" rel="nofoll...   
4  1005232370834979968  <a href="http://www.echosign.com/" rel="nofoll...   

                       created_at lang  verified  user_id_str  geo_enabled  \
0  Tue Jun 05 23:48:44 +0000 2018   en     False    370896174         True   
1  Fri Jun 08 21:59:44 +0000 2018   en     False    589580923        False   
2  Sat Jun 09 21:32:39 +0000 2018   en     False     18226845        False   
3  Thu Jun 07 22:43:56 +0000 2018   en     False    745859413        False   
4  Fri Jun 08 23:37:28 +0000 2018   en     False    183367252        False   

                  user_created_at  
0  Fri Sep 09 20:40:00 +0000 2011  
1  Fri May 25 01:27:15 +0000 2012  
2  Thu Dec 18 22:06:40 +0000 2008  
3  Wed Aug 08 20:11:12 +0000 2012  
4  Thu Aug 26 20:50:06 +0000 2010

Following the advice from Matt Drury, the one of data scientists I admire most, I will start small - starting with 10 columns.

I want to start with

1) text: characteristic about post, will be TF-IDF

2) description: characteristic about account, will be TF-IDF

3) statuses/followers/favourites/friends/listed count: all measure the activity of the count (good numeric values as well - to identify account identity (too high vs. too low)

4) company boolean category (DocuSign, OneSpan, SignNow, AdobeSign): if certain company is likely to focus marketing / easy to highjack? 

5) created at: post created time would be a good predictor (if spam is likely to be posted on some time range)


Now, I will continue cleaning data.

I will remove rows with non-English and all remaining column I think which are irrelevant/less important for making a decision. 

In [34]:
df3 = df2[df2['lang']=='en']

In [35]:
df3['lang'].unique()

array(['en'], dtype=object)

In [36]:
df3.columns

Index(['index', 'text', 'spam', 'spam_marketing', 'spam_hijack',
       'spam_corporate', 'spam_bot', 'spam_known', 'spam_own', 'Docusign',
       'onespan', 'signnow', 'adobe sign', 'id_str', 'source', 'created_at',
       'lang', 'listed_count', 'verified', 'location', 'user_id_str',
       'description', 'geo_enabled', 'user_created_at', 'statuses_count',
       'followers_count', 'favourites_count', 'name', 'user_lang',
       'friends_count', 'screen_name', 'expanded_url', 'url'],
      dtype='object')

In [38]:
columns_to_drop = ['index','id_str','source','lang','verified',
                   'location','user_id_str','geo_enabled','user_created_at',
                   'name','user_lang','screen_name','expanded_url','url']

In [39]:
df3.drop(columns_to_drop, axis=1,inplace=True)

/home/daniel/anaconda3/lib/python3.6/site-packages/pandas/core/frame.py:3940: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


In [41]:
df3

text  spam  \
0      just closed a deal in 27 hours using #AdobeSig...     0   
1      just closed a deal in 2 days using #AdobeSign ...     0   
2      just closed a deal in 2 hours using #AdobeSign...     0   
3      just closed a deal in 26 hours using #AdobeSig...     0   
4      just closed a deal in 6 days using #AdobeSign ...     0   
5      just closed a deal in 35 minutes using #AdobeS...     0   
6      just closed a deal in 3 days using #AdobeSign ...     0   
7      just closed a deal in 58 minutes using #AdobeS...     0   
8      just closed a deal in 7 minutes using #AdobeSi...     0   
9      So apparently IU is catching on and won’t ju...     0   
10     just closed a deal in 27 minutes using #AdobeS...     0   
11     just closed a deal in 21 minutes using #AdobeS...     0   
12     just closed a deal in 4 hours using #AdobeSign...     0   
13     just closed a deal in 28 minutes using #AdobeS...     0   
14     just closed a deal using #AdobeSign https://t....     0   
15     just closed a deal in 25 hours using #AdobeSig...     0   
16     just closed a deal in 3 hours using #AdobeSign...     0   
17     just closed a deal in 8 minutes using #AdobeSi...     0   
18     just closed a deal in 88 minutes using #AdobeS...     0   
19     just closed a deal in 29 hours using #AdobeSig...     0   
20     just closed a deal in 18 minutes using #AdobeS...     0   
21     just closed a deal in 5 minutes using #AdobeSi...     0   
22     just signed an agreement using #AdobeSign http...     0   
23     just closed a deal in 3 minutes using #AdobeSi...     0   
24     just closed a deal in 75 minutes using #AdobeS...     0   
25     just closed a deal in 4 minutes using #AdobeSi...     0   
26     just closed a deal in 12 minutes using #AdobeS...     0   
27     just closed a deal in 73 minutes using #AdobeS...     0   
28     just closed a deal in 3 hours using #AdobeSign...     0   
29     just closed a deal in 5 days using #AdobeSign ...     0   
...                                                  ...   ...   
19981  SignNow rescues your company from the unnecess...     0   
19982  If you send out 1000 docs for HR/Comp signing,...     0   
19983  Extend that last mile with a one-way or two-wa...     0   
19984  SignNow helps put the puzzle pieces together f...     0   
19985  With #SignNow, you are flexible enough to chan...     0   
19986  Do you want a digital interactive &amp; engagi...     0   
19987  @atharkhan @HelloSign @G2Crowd @RightSignature...     0   
19988  Are you looking to make ACH Payments using Str...     0   
19989  5 major benefits of choosing SignNow for #medi...     0   
19990  Do you need to send a document to a group of p...     0   
19991  @beezy It should be back up. It was down for a...     0   
19992  #Didyouknow that you can create a link to put ...     0   
19993  SignNow helps get patient forms signed faster ...     0   
19994  You probably knew that you can save money with...     0   
19995  Need to sign last minute #legal forms? SignNow...     0   
19996  @claaack Hi Mr. November, I apologize. We are ...     0   
19997  and sign the petition as it needs lots &amp; l...     0   
19998  Using an Electronic Signature or E-Signature a...     0   
19999  #LaborDay Weekend yet? Don't let work hold you...     0   
20000  Are you having an event for your office for th...     0   
20001  Summer is fading but the dollars in your bank ...     0   
20002  Do you need to sign documents in person on an ...     0   
20003  Are you stressed about getting a document sign...     0   
20004  Please join in the effort to stop genocide in ...     0   
20005  @Anafa It should be back up. It was down for a...     0   
20006  @signnow Yes, they apologized for the issues b...     0   
20007  SignNow ensures that you cannot be breached by...     0   
20008  @claaack Hi Mr. November, Were you able to get...     0   
20009  Want extra money this #LaborDay weekend withou...     0   
20010  Are you in #HR? Sav

20006 rows (seems like 5 rows were written with foreign languages) and 19 columns.

Here, 7 columns are y-value (spam related columns)

I will manipulate 'created_at' column first. I will use (Matt's method) to create two additional columns from it.

I only need hour:minute:second, and convert it to decimal form.

In [49]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20006 entries, 0 to 20010
Data columns (total 19 columns):
text                20006 non-null object
spam                20006 non-null int64
spam_marketing      20006 non-null int64
spam_hijack         20006 non-null int64
spam_corporate      20006 non-null int64
spam_bot            20006 non-null int64
spam_known          20006 non-null int64
spam_own            20006 non-null int64
Docusign            20006 non-null bool
onespan             20006 non-null bool
signnow             20006 non-null bool
adobe sign          20006 non-null bool
created_at          20006 non-null object
listed_count        20006 non-null int64
description         14781 non-null object
statuses_count      20006 non-null int64
followers_count     20006 non-null int64
favourites_count    20006 non-null int64
friends_count       20006 non-null int64
dtypes: bool(4), int64(12), object(3)
memory usage: 2.5+ MB


In [78]:
df3['created_at'].head()

0    Tue Jun 05 23:48:44 +0000 2018
1    Fri Jun 08 21:59:44 +0000 2018
2    Sat Jun 09 21:32:39 +0000 2018
3    Thu Jun 07 22:43:56 +0000 2018
4    Fri Jun 08 23:37:28 +0000 2018
Name: created_at, dtype: object

take out hh:mm:ss, convert it to decimal number...

In [62]:
df3['time_created'] = pd.to_datetime(df3['created_at']).copy()

/home/daniel/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [63]:
df3.head()

text  spam  spam_marketing  \
0  just closed a deal in 27 hours using #AdobeSig...     0               0   
1  just closed a deal in 2 days using #AdobeSign ...     0               0   
2  just closed a deal in 2 hours using #AdobeSign...     0               0   
3  just closed a deal in 26 hours using #AdobeSig...     0               0   
4  just closed a deal in 6 days using #AdobeSign ...     0               0   

   spam_hijack  spam_corporate  spam_bot  spam_known  spam_own  Docusign  \
0            0               0         1           0         0     False   
1            0               0         1           0         0     False   
2            0               0         1           0         0     False   
3            0               0         1           0         0     False   
4            0               0         1           0         0     False   

   onespan  signnow  adobe sign                      created_at  listed_count  \
0    False    False        True  Tue Jun 05 23:48:44 +0000 2018             1   
1    False    False        True  Fri Jun 08 21:59:44 +0000 2018            14   
2    False    False        True  Sat Jun 09 21:32:39 +0000 2018             0   
3    False    False        True  Thu Jun 07 22:43:56 +0000 2018             2   
4    False    False        True  Fri Jun 08 23:37:28 +0000 2018             8   

                                         description  statuses_count  \
0  There's no substitute for experience! Call 609...            6750   
1  We provide an affordable way to have a wonderf...            5018   
2                                                NaN              29   
3                                                NaN            5332   
4  Carefully Crafted & Intentionally Inspired Eve...            1337   

   followers_count  favourites_count  friends_count              time_created  
0              108                 1             12 2018-06-05 23:48:44+00:00  
1               47                 2             25 2018-06-08 21:59:44+00:00  
2               60                10             52 2018-06-09 21:32:39+00:00  
3               66                 0            182 2018-06-07 22:43:56+00:00  
4              225                19            348 2018-06-08 23:37:28+00:00

In [79]:
df3['time_hms_only'] = df3['time_created'].dt.time
#unnecessary - I can work with time_created

/home/daniel/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [76]:
df3['created_at_float'] = df3['time_created'].dt.hour + df3['time_created'].dt.minute/60 + df3['time_created'].dt.second/3600

/home/daniel/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [80]:
df4 = df3.drop(['created_at','time_hms_only','time_created'],axis=1)

In [81]:
df4.head()

text  spam  spam_marketing  \
0  just closed a deal in 27 hours using #AdobeSig...     0               0   
1  just closed a deal in 2 days using #AdobeSign ...     0               0   
2  just closed a deal in 2 hours using #AdobeSign...     0               0   
3  just closed a deal in 26 hours using #AdobeSig...     0               0   
4  just closed a deal in 6 days using #AdobeSign ...     0               0   

   spam_hijack  spam_corporate  spam_bot  spam_known  spam_own  Docusign  \
0            0               0         1           0         0     False   
1            0               0         1           0         0     False   
2            0               0         1           0         0     False   
3            0               0         1           0         0     False   
4            0               0         1           0         0     False   

   onespan  signnow  adobe sign  listed_count  \
0    False    False        True             1   
1    False    False        True            14   
2    False    False        True             0   
3    False    False        True             2   
4    False    False        True             8   

                                         description  statuses_count  \
0  There's no substitute for experience! Call 609...            6750   
1  We provide an affordable way to have a wonderf...            5018   
2                                                NaN              29   
3                                                NaN            5332   
4  Carefully Crafted & Intentionally Inspired Eve...            1337   

   followers_count  favourites_count  friends_count  created_at_float  
0              108                 1             12         23.812222  
1               47                 2             25         21.995556  
2               60                10             52         21.544167  
3               66                 0            182         22.732222  
4              225                19            348         23.624444

Taking the advice from Jack (and Matt who told me the name of this method), I converted time with fourier transformation.

In [84]:
df4['time_float_sin'] = np.sin((df4['created_at_float']*2*np.pi)/24)

In [85]:
df4['time_float_cos'] = np.cos((df4['created_at_float']*2*np.pi)/24)

In [87]:
df4.drop('created_at_float',axis=1,inplace=True)

In [88]:
df4.head()

text  spam  spam_marketing  \
0  just closed a deal in 27 hours using #AdobeSig...     0               0   
1  just closed a deal in 2 days using #AdobeSign ...     0               0   
2  just closed a deal in 2 hours using #AdobeSign...     0               0   
3  just closed a deal in 26 hours using #AdobeSig...     0               0   
4  just closed a deal in 6 days using #AdobeSign ...     0               0   

   spam_hijack  spam_corporate  spam_bot  spam_known  spam_own  Docusign  \
0            0               0         1           0         0     False   
1            0               0         1           0         0     False   
2            0               0         1           0         0     False   
3            0               0         1           0         0     False   
4            0               0         1           0         0     False   

   onespan  signnow  adobe sign  listed_count  \
0    False    False        True             1   
1    False    False        True            14   
2    False    False        True             0   
3    False    False        True             2   
4    False    False        True             8   

                                         description  statuses_count  \
0  There's no substitute for experience! Call 609...            6750   
1  We provide an affordable way to have a wonderf...            5018   
2                                                NaN              29   
3                                                NaN            5332   
4  Carefully Crafted & Intentionally Inspired Eve...            1337   

   followers_count  favourites_count  friends_count  time_float_sin  \
0              108                 1             12       -0.049140   
1               47                 2             25       -0.501007   
2               60                10             52       -0.599548   
3               66                 0            182       -0.325843   
4              225                19            348       -0.098162   

   time_float_cos  
0        0.998792  
1        0.865443  
2        0.800339  
3        0.945424  
4        0.995170